internal/webhooks/machinedeployment.go

```diff
@@ -36,40 +36,39 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/cluster-api/feature"
 	"sigs.k8s.io/cluster-api/util/version"
 )
 
-func (m *MachineDeployment) SetupWebhookWithManager(mgr ctrl.Manager) error {
+func (webhook *MachineDeployment) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	if webhook.Decoder == nil {
+		webhook.Decoder = admission.NewDecoder(mgr.GetScheme())
+	}
+
 	// This registers MachineDeployment as a validating webhook and
-	// machineDeploymentDefaulter as a defaulting webhook.
+	// MachineDeploymentDefaulter as a defaulting webhook.
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(m).
-		WithDefaulter(MachineDeploymentDefaulter(mgr.GetScheme())).
+		For(&clusterv1.MachineDeployment{}).
+		WithDefaulter(webhook).
+		WithValidator(webhook).
 		Complete()
 }
 
 // +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1beta1-machinedeployment,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=machinedeployments,versions=v1beta1,name=validation.machinedeployment.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
 // +kubebuilder:webhook:verbs=create;update,path=/mutate-cluster-x-k8s-io-v1beta1-machinedeployment,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=machinedeployments,versions=v1beta1,name=default.machinedeployment.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
 
-var _ webhook.CustomDefaulter = &machineDeploymentDefaulter{}
-var _ webhook.Validator = &MachineDeployment{}
-
-// MachineDeploymentDefaulter creates a new CustomDefaulter for MachineDeployments.
-func MachineDeploymentDefaulter(scheme *runtime.Scheme) webhook.CustomDefaulter {
-	return &machineDeploymentDefaulter{
-		decoder: admission.NewDecoder(scheme),
-	}
+// MachineDeployment implements a validation and defaulting webhook for MachineDeployment.
+type MachineDeployment struct {
+	Decoder *admission.Decoder
 }
 
-// machineDeploymentDefaulter implements a defaulting webhook for MachineDeployment.
-type machineDeploymentDefaulter struct {
-	decoder *admission.Decoder
-}
+var _ webhook.CustomDefaulter = &MachineDeployment{}
+var _ webhook.CustomValidator = &MachineDeployment{}
 
 // Default implements webhook.CustomDefaulter.
-func (webhook *machineDeploymentDefaulter) Default(ctx context.Context, obj runtime.Object) error {
-	m, ok := obj.(*MachineDeployment)
+func (webhook MachineDeployment) Default(ctx context.Context, obj runtime.Object) error {
```

Comment:
```suggestion
func (webhook *MachineDeployment) Default(ctx context.Context, obj runtime.Object) error {
```
same below

Reply:
Ahh no wait, my bad, let me sort that

Reply:
Just got a warning from Intellij because it doesn't like it if some methods use a value and some a pointer receiver

Reply:
Should be fixed on latest push